public/components/model/model_tags_panel/model_tag_key_field.tsx

```diff
@@ -0,0 +1,245 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import React, { useCallback, useMemo } from 'react';
+import {
+  EuiButtonIcon,
+  EuiComboBox,
+  EuiComboBoxOptionOption,
+  EuiContext,
+  EuiFieldText,
+  EuiFlexGroup,
+  EuiFlexItem,
+  EuiFormRow,
+  EuiText,
+  EuiToken,
+} from '@elastic/eui';
+import { Control, useController, useWatch } from 'react-hook-form';
+
+import { TagKeyFormData } from '../types';
+import { TagKey, tagKeyTypeOptions } from '../../common';
+
+const FORM_ITEM_WIDTH = 400;
+
+const MAX_TAG_LENGTH = 80;
+
+const KEY_COMBOBOX_I18N = {
+  mapping: {
+    'euiComboBoxOptionsList.noAvailableOptions': 'No keys found. Add a key.',
+  },
+};
+
+const singleSelection = { asPlainText: true };
+
+interface ModelTagKeyFieldProps {
+  control: Control<TagKeyFormData>;
+  index: number;
+  onRemove: (index: number) => void;
+  allTagKeys: TagKey[];
+  savedTagKeys: TagKey[];
+  showLabel: boolean;
+}
+
+export const ModelTagKeyField = ({
+  control,
+  index,
+  onRemove,
+  allTagKeys,
+  showLabel,
+  savedTagKeys,
+}: ModelTagKeyFieldProps) => {
+  const tagKeysInForm = useWatch({
+    name: 'tagKeys',
+    control,
+  });
+  const nameController = useController({
+    name: `tagKeys.${index}.name`,
+    control,
+    rules: {
+      maxLength: {
+        value: MAX_TAG_LENGTH,
+        message: '80 characters allowed. Use 80 characters or less.',
+      },
+      validate: (tagKey) => {
+        if (tagKeysInForm) {
+          // If a tag has key, validate if the same tag key was added before
+          if (tagKey) {
+            // Find if tag key already exists in saved tag key list, this is a rare case may caused by the saved tag keys updated
+            for (let i = 0; i < savedTagKeys.length; i++) {
+              if (savedTagKeys[i].name === tagKey) {
+                return 'Tag keys must be unique. Use a unique key.';
+              }
+            }
+            // Find if the same tag key appears before the current tag key
+            for (let i = 0; i < index; i++) {
+              // If found the same tag key, then the current tag key is invalid
+              if (tagKeysInForm[i].name === tagKey) {
+                return 'Tag keys must be unique. Use a unique key.';
+              }
+            }
+          }
+        }
+        return true;
+      },
+    },
+  });
+
+  const typeController = useController({
+    name: `tagKeys.${index}.type`,
+    control,
+  });
+
+  const { ref: nameFieldRef, ...restNameFieldProps } = nameController.field;
+  const { ref: typeFieldRef, ...restTypeFieldProps } = typeController.field;
+
+  const keyOptions = useMemo(() => {
+    const savedTagKeyMap = savedTagKeys.reduce<{ [key: string]: boolean }>(
+      (pValue, cValue) => ({
+        ...pValue,
+        [`${cValue.name}${cValue.type}`]: true,
+      }),
+      {}
+    );
+    return allTagKeys.map((item) => ({
+      label: item.name,
+      value: item,
+      disabled: savedTagKeyMap[`${item.name}${item.type}`],
```

Comment:
Q: should it just check `item.name`? because adding names with different type is not allowed, am I right?

Reply:
Yes. I think we can check by `item.name`. Ideally one tag name only have one type. Use name check if more safe.